src/libs/elektra/key.c

```diff
@@ -244,218 +244,220 @@ Key * keyVNew (const char * name, va_list va)
+
+	if (source == NULL)
 	{
-		keyClear (dest);
-		return 0;
+		if (test_bit (flags, KEY_CP_NAME))
+		{
+			keySetName (dest, "/");
+		}
+		if (test_bit (flags, KEY_CP_VALUE))
+		{
+			keySetRaw (dest, NULL, 0);
+		}
+		if (test_bit (flags, KEY_CP_META))
+		{
+			ksClear (dest->meta);
+		}
+		return dest;
 	}
 
-	// remember dynamic memory to be removed
-	char * destKey = dest->key;
-	char * destUKey = dest->ukey;
-	void * destData = dest->data.c;
-	KeySet * destMeta = dest->meta;
+	if (test_bit (flags, KEY_CP_STRING) && keyIsBinary (source)) return NULL;
 
-	// duplicate dynamic properties
-	if (source->key)
-	{
-		dest->key = elektraStrNDup (source->key, source->keySize);
-		if (!dest->key) goto memerror;
-	}
-	else
-	{
-		dest->key = 0;
-	}
+	if (source == dest) return dest;
 
-	if (source->ukey)
-	{
-		dest->ukey = elektraStrNDup (source->ukey, source->keyUSize);
-		if (!dest->ukey) goto memerror;
-	}
-	else
-	{
-		dest->ukey = 0;
-	}
+	// remember original data of dest
+	Key orig = *dest;
 
+	// TODO: check MMAP flags
 
-	if (source->data.v)
+	// duplicate dynamic properties
+	if (test_bit (flags, KEY_CP_NAME))
 	{
-		dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
-		if (!dest->data.v) goto memerror;
+		if (source->key)
+		{
+			dest->key = elektraStrNDup (source->key, source->keySize);
+			if (!dest->key) goto memerror;
+			dest->keySize = source->keySize;
+
+			ELEKTRA_ASSERT (source->ukey != NULL, "key != NULL but ukey == NULL");
+			dest->ukey = elektraStrNDup (source->ukey, source->keyUSize);
+			if (!dest->ukey) goto memerror;
+			dest->keyUSize = source->keyUSize;
+		}
+		else
+		{
+			dest->key = elektraStrDup ("/");
+			dest->keySize = 2;
+
+			dest->ukey = elektraMalloc (3);
+			dest->ukey[0] = KEY_NS_CASCADING;
+			dest->ukey[1] = '\0';
+			dest->ukey[2] = '\0';
+			dest->keyUSize = 3;
+		}
+		clear_bit (dest->flags, KEY_FLAG_MMAP_KEY);
 	}
-	else
+
+	if (test_bit (flags, KEY_CP_STRING))
 	{
-		dest->data.v = 0;
+		if (source->data.v)
+		{
+			dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
+			if (!dest->data.v) goto memerror;
+			dest->dataSize = source->dataSize;
+
+			if (!test_bit (flags, KEY_META) && keyIsBinary (source))
+			{
+				keySetMeta (dest, "binary", "");
+			}
+		}
+		else
+		{
+			dest->data.v = NULL;
+			dest->dataSize = 0;
+		}
+		clear_bit (dest->flags, KEY_FLAG_MMAP_DATA);
 	}
 
-	if (source->meta)
+	if (test_bit (flags, KEY_CP_VALUE))
 	{
-		dest->meta = ksDup (source->meta);
-		if (!dest->meta) goto memerror;
+		if (source->data.v)
+		{
+			dest->data.v = elektraStrNDup (source->data.v, source->dataSize);
+			if (!dest->data.v) goto memerror;
+			dest->dataSize = source->dataSize;
+
+			if (!test_bit (flags, KEY_META) && keyIsBinary (source))
+			{
+				keySetMeta (dest, "binary", "");
+			}
+		}
+		else
+		{
+			dest->data.v = NULL;
+			dest->dataSize = 0;
+		}
+		clear_bit (dest->flags, KEY_FLAG_MMAP_DATA);
```

Comment:
All of these look good to me.

It's simple: any time we replace a pointer to an mmap region with a pointer returned by malloc() we need to reset the flag to avoid memleaks. Unfortunately we may not call free() on anything in the mmap region.